lib/rules/no-shadow.js

```diff
@@ -109,18 +124,24 @@ module.exports = function(context) {
      * Checks if the given variables are shadowed in the given scope.
      * @param {Array} variables The variables to look for
      * @param {Object} scope The scope to be checked.
-     * @returns {void}
+     * @returns {Array} Variables which are already declared in the given scope.
      */
     function checkShadowsInScope(variables, scope) {
+
+        var shadowVars = [];
```

Comment:
Might be faster to use a hash, e.g.:

```js
function checkShadowsInScope(variables, scope, reported) {
    //snip...
    if (!reported[variable.name]){
        reported[variable.name] = true;
        context.report(...);
    }
    //snip...
}
```
And then calling this via:
```js
var reported = Object.create(null);
while (upper) {
    checkShadowsInScope(variables, upper, reported);
    upper = upper.upper.
}
```

You'd probably need to attach the location to the `variable.name` value to ensure that you're reporting each location, but you're not diffing two arrays in each scope.